Added bezier curve shape type

leins275 opened the pull request:
Added ability to create shapes made of bezier curve segments.

leins275 committed:
Added menu option for curve creation

leins275 committed:
Added curve menu option functionality

leins275 committed:
Added curve option to lists in createMode functions

leins275 committed:
In progress
 - creating support for drawing curves

leins275 committed:
In progress
 - modifying addPoint function

leins275 committed:
Updated addPoint function

leins275 committed:
In progress

 - fixed some bugs with curves

leins275 committed:
Fixed problem with point type

leins275 committed:
Fixed drawVertex func

leins275 committed:
Fixed
 - first curve segment
 - all problems with adding curve points issues

leins275 committed:
Added
 - draw support dash lines between curve points

leins275 committed:
Changed
 - curve support lines color
Added
 - possibility to close curve line

leins275 committed:
Added
 - support for end with bezier line segment

leins275 committed:
Updated
 - curve object filling
 - source curve lines type

leins275 committed:
Fixed
 - problem with line_path.simplified
   (should work only with curves)

leins275 committed:
Fixed
 - curve object filling while drawing

leins275 committed:
In progress
 - updating canvas line shape processing for curves.

leins275 committed:
In progress
 - refactored canvas.py

leins275 committed:
Added
 - drawing source lines correctly

leins275 committed:
Updated
 - now canvas.line processing is correct

leins275 committed:
Added
 - ability to save/load curve objects

Commit:
Added
 - blocking ability to add points out of canvas

Commit:
Fixed
 - don't draw useless support lines

Commit:
Updated
 - deprecated draw points out of canvas(improved)

Commit:
Fixed
 - repaint canvas.line shape error

Commit:
Updated
 - fixed small troubles

daniilsavchuk commented on labelme/app.py:
```diff
@@ -788,41 +802,55 @@ def toggleDrawMode(self, edit=True, createMode='polygon'):
+                self.actions.createLineMode.setEnabled(True)
+                self.actions.createPointMode.setEnabled(True)
+                self.actions.createLineStripMode.setEnabled(True)
+                self.actions.createCurveMode.setEnabled(False)
```
Why code is duplicated? Could we trim it and parametrize? Should be shortened even on the labelme side

leins275 replied:
Refactored.

daniilsavchuk commented on labelme/config/default_config.yaml:
```diff
@@ -68,6 +68,7 @@ shortcuts:
   create_line: null
   create_point: null
   create_linestrip: null
+  create_curve: null
```
may we use ctrl+B for the curve?

leins275 replied:
Yes, done.

daniilsavchuk commented on labelme/shape.py:
```diff
@@ -40,6 +42,8 @@ def __init__(self, label=None, line_color=None, shape_type=None,
         self.label = label
         self.points = []
+        self.segments = []
+        self.segments_len = 0
```
cannot we determine segment length getting len(segments) ?

leins275 replied:
It has different semantics. I renamed this var, please check commit messages.

daniilsavchuk commented on labelme/shape.py:
```diff
@@ -71,18 +75,67 @@ def shape_type(self, value):
+        new_segment = [seg_begin, seg_len]
+        self.segments.append(new_segment)
+
+    # TODO:
```
No todos in productive!

leins275 replied:
deleted

daniilsavchuk commented on labelme/shape.py:
```diff
@@ -71,18 +75,67 @@ def shape_type(self, value):
+            seg_begin = len(self.points) - size
+            seg_len = size
+        else:
+            size = 3 + degree_increment
```
2 and 3 are magic numbersd

leins275 replied:
refactored

daniilsavchuk commented on labelme/widgets/canvas.py:
```diff
@@ -148,6 +148,85 @@ def unHighlight(self):
+            if edgeY >= size.height() - 1:
+                edgeY = size.height() - 1
+                edgeX = (xa - xb) / (ya - yb) * edgeY + (- yb * xa + xb * ya) / (ya - yb)
+        return edgeX, edgeY
```
Could be essentially improved if using min max and storing precomputed values

LanskovNV replied:
done

leins275 replied:
done.

leins275 committed:
Merge branch 'master' into bezier-shape

daniilsavchuk reviewed (commented):
Please perform improvements

leins275 committed:
Refactored
 - refactored dublicated code in app.py by adding setCreateActions func

leins275 committed:
Added
 - new hotkey [Ctrl+B] to enable create curve mode

leins275 committed:
Refactored
 - renamed segments_len in shape.py to points_in_segments. This variable stores num of points from self.points that belong to segment.

leins275 committed:
Deleted TODO item in shape.py

leins275 committed:
Refactored
 - reduced some magic numbers from shape.py

leins275 committed:
Updated
 - improved updatePixmapPoints func in canvas.py

leins275 committed:
Fixed

daniilsavchuk commented on labelme/shape.py:
```diff
@@ -71,18 +77,65 @@ def shape_type(self, value):
+                        self.points.pop(-1)
+                        self.close()
+                except NoPointError as e:
+                    assert False, "new_p is empty !!!"
```
Asserts are debugging feature only, so throw an exception forward and just log an error message

daniilsavchuk commented on labelme/widgets/canvas.py:
```diff
@@ -155,6 +155,78 @@ def unHighlight(self):
+
+        if xa - xb != 0:
+            edgeX = min(max_width, max(edgeX, 0))
+            edgeY = (ya - yb) / (xa - xb) * edgeX + (- xb * ya + yb * xa) / (xa - xb)
```
please define operations you are using like coefficients. Now it is code duplication here

leins275 replied:
done

daniilsavchuk commented on labelme/widgets/canvas.py:
```diff
@@ -155,6 +155,78 @@ def unHighlight(self):
+            edgeY = (ya - yb) / (xa - xb) * edgeX + (- xb * ya + yb * xa) / (xa - xb)
+        if ya - yb != 0:
+            edgeY = min(max_height, max(edgeY, 0))
+            edgeX = (xa - xb) / (ya - yb) * edgeY + (- yb * xa + xb * ya) / (ya - yb)
```
And here

leins275 replied:
done

daniilsavchuk reviewed (changes requested):
Should do other fixes before merging, however it looks almost perfect right now

leins275 committed:
Refactored dublicate code fragment
